AppCenter/AppCenter/Internals/Model/Util/MSAppCenterUserDefaults.m

```diff
@@ -0,0 +1,135 @@
+
+@implementation MSAppCenterUserDefaults
+
++ (instancetype)shared {
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [[MSAppCenterUserDefaults alloc] init];
+    NSDictionary *migratedKeys = @{
+      @"MSChannelStartTimer" : @"MSACChannelStartTimer",
+      @"pastDevicesKey" : @"MSACPastDevicesKey",
+      @"MSInstallId" : @"MSACInstallId",
+      @"MSAppCenterIsEnabled" : @"MSACAppCenterIsEnabled",
+      @"MSEncryptionKeyMetadata" : @"MSACEncryptionKeyMetadata",
+      @"SessionIdHistory" : @"MSACSessionIdHistory",
+      @"UserIdHistory" : @"MSACUserIdHistory"
+    };
+    [sharedInstance migrateKeys:migratedKeys forService:@"Core"];
+  });
+  return sharedInstance;
+}
+
++ (void)resetSharedInstance {
+  onceToken = 0; // resets the once_token so dispatch_once will run again
+  sharedInstance = nil;
+}
+
+- (void)migrateKeys:(NSDictionary *)migratedKeys forService:(NSString *)serviceName {
+  NSString *serviceHasMigratedKey = [NSString stringWithFormat:kMSAppCenterUserDefaultsMigratedKeyFormat, serviceName];
+  NSNumber *serviceHasMigrated = [self objectForKey:serviceHasMigratedKey];
+  if (serviceHasMigrated) {
+    return;
+  }
+  MSLogVerbose([MSAppCenter logTag], @"Migrating the old NSDefaults keys to new ones.");
+  for (NSString *oldKey in [migratedKeys allKeys]) {
+    NSString *newKey = migratedKeys[oldKey];
+    id value = [[NSUserDefaults standardUserDefaults] objectForKey:oldKey];
+    if (value != nil) {
+      [[NSUserDefaults standardUserDefaults] setObject:value forKey:newKey];
+      [[NSUserDefaults standardUserDefaults] removeObjectForKey:oldKey];
+      MSLogVerbose([MSAppCenter logTag], @"Migrating the key %@ -> %@", oldKey, newKey);
+    }
+  }
+  [self setObject:@(1) forKey:serviceHasMigratedKey];
+}
+
+- (NSString *)getAppCenterKeyFrom:(NSString *)key {
+    if ([key hasPrefix:kMSUserDefaultsPrefix]) {
+        return key;
+    } else {
+        return [kMSUserDefaultsPrefix stringByAppendingString:key];
+    }
+}
+
+- (id)objectForKey:(NSString *)key {
+  NSString *keyPrefixed = [self getAppCenterKeyFrom:key];
+  return [[NSUserDefaults standardUserDefaults] objectForKey:keyPrefixed];
+}
+
+- (void)setObject:(id)o forKey:(NSString *)key {
+  NSString *keyPrefixed = [self getAppCenterKeyFrom:key];
+  [[NSUserDefaults standardUserDefaults] setObject:o forKey:keyPrefixed];
+}
+
+- (void)removeObjectForKey:(NSString *)key {
+  NSString *keyPrefixed = [self getAppCenterKeyFrom:key];
+  [[NSUserDefaults standardUserDefaults] removeObjectForKey:keyPrefixed];
+}
+
+- (NSDictionary *)updateDictionary:(NSDictionary *)dict forKey:(NSString *)key expiration:(float)expiration {
+  NSString *keyPrefixed = [self getAppCenterKeyFrom:key];
+  NSMutableDictionary *update = [[NSMutableDictionary alloc] initWithDictionary:dict];
+
+  // Get from local store.
+  NSDictionary *store = [[NSUserDefaults standardUserDefaults] dictionaryForKey:keyPrefixed];
+
+  CFAbsoluteTime ts = [(NSNumber *)store[kMSUserDefaultsTs] doubleValue];
```

Comment:
What is ts? Use full name and add a comment

Reply:
I'm not sure what it is exactly... 🤔

Reply:
That was for values expiration but I don't this we use it.